systemd-boot relies on wchar_t being __WCHAR_TYPE__ but musl sets it as
int here: https://git.musl-libc.org/cgit/musl/tree/arch/x86_64/bits/alltypes.h.in?h=v1.2.4#n9

So the patch below sets __DEFINED_wchar_t and loads `stddef.h` later so that we can override it.

diff --git src/boot/efi.h src/boot/efi.h
index b8d193154e..e432e9c91f 100644
--- src/boot/efi.h
+++ src/boot/efi.h
@@ -3,14 +3,15 @@

 #include <stdarg.h>
 #include <stdbool.h>
-#include <stddef.h>
 #include <stdint.h>

 #include "macro-fundamental.h"

 #if SD_BOOT
 /* uchar.h/wchar.h are not suitable for freestanding environments. */
+#define __DEFINED_wchar_t
 typedef __WCHAR_TYPE__ wchar_t;
+#include <stddef.h>
 typedef __CHAR16_TYPE__ char16_t;
 typedef __CHAR32_TYPE__ char32_t;

diff --git src/fundamental/macro-fundamental.h src/fundamental/macro-fundamental.h
index 1d49765fce..113bda78c7 100644
--- src/fundamental/macro-fundamental.h
+++ src/fundamental/macro-fundamental.h
@@ -3,6 +3,9 @@

 #if !SD_BOOT
 #  include <assert.h>
+#else
+#define __DEFINED_wchar_t
+#include <stddef.h>
 #endif

 #include <limits.h>
diff --git src/fundamental/memory-util-fundamental.h src/fundamental/memory-util-fundamental.h
index 78e2dbec59..4f68b56e9d 100644
--- src/fundamental/memory-util-fundamental.h
+++ src/fundamental/memory-util-fundamental.h
@@ -1,10 +1,10 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 #pragma once

-#include <stddef.h>
-
 #if SD_BOOT
 #  include "efi-string.h"
+#define __DEFINED_wchar_t
+#include <stddef.h>
 #else
 #  include <string.h>
 #endif
